builtin/logical/transit/path_keys.go

```diff
@@ -13,12 +15,42 @@ type Policy struct {
 	Name       string `json:"name"`
 	Key        []byte `json:"key"`
 	CipherMode string `json:"cipher"`
+
+	// Derived keys MUST provide a context and the
+	// master underlying key is never used.
+	Derived bool   `json:"derived"`
+	KDFMode string `json:"kdf_mode"`
 }
 
 func (p *Policy) Serialize() ([]byte, error) {
 	return json.Marshal(p)
 }
 
+// DeriveKey is used to derive the encryption key that should
+// be used depending on the policy. If derivation is disabled the
+// raw key is used and no context is required, otherwise the KDF
+// mode is used with the context to derive the proper key.
+func (p *Policy) DeriveKey(context []byte) ([]byte, error) {
+	// Fast-path non-derived keys
+	if !p.Derived {
+		return p.Key, nil
+	}
+
+	// Ensure a context is provided
+	if len(context) == 0 {
+		return nil, fmt.Errorf("missing context for key derivation")
```

Comment:
I think we could improve the UX here a bit, especially since the person who created the key with derivation may not be the person who is writing to it. Maybe something like:

```
missing 'context' for key deriviation. The key {p.Name} was created using a derived key, which means additional, per-request information must be included in order to encrypt or decrypt information. Please see the Vault documentation at ... for more information.
```